### ISSUE TITLE
sql: small LIMIT value causes results to be omitted

### ISSUE THREAD

**Issue by jrupac** [cockroachdb/cockroach#16313]
1. Please supply the header (i.e. the first few lines) of your most recent
   log file **for each node in your cluster**. On most unix-based systems
   running with defaults, this boils down to the output of

     grep -F '[config]' cockroach-data/logs/cockroach.log

   When log files are not available, supply the output of `cockroach version`
   and all flags/environment variables passed to `cockroach start` instead.

```shell
$ grep -F '[config]' cockroach-data/logs/cockroach.log
I170605 02:04:31.905759 1 util/log/clog.go:1011  [config] file created at: 2017/06/05 02:04:31
I170605 02:04:31.905759 1 util/log/clog.go:1011  [config] running on machine: samurai
I170605 02:04:31.905759 1 util/log/clog.go:1011  [config] binary: CockroachDB CCL v1.0.1 (linux amd64, built 2017/05/25 15:17:49, go1.8.1)
I170605 02:04:31.905759 1 util/log/clog.go:1011  [config] arguments: [cockroach start --insecure]
```

2. Please describe the issue you observed:

- What did you do?

Here's my table schema:

```sql
root@:26257/Goliath> SHOW CREATE TABLE article;
+---------+------------------------------------------------------------------------------------------------------+
|  Table  |                                             CreateTable                                              |
+---------+------------------------------------------------------------------------------------------------------+
| article | CREATE TABLE article (␤                                                                              |
|         |     id INT NOT NULL DEFAULT unique_rowid(),␤                                                         |
|         |     feed INT NOT NULL,␤                                                                              |
|         |     folder INT NOT NULL,␤                                                                            |
|         |     hash STRING NULL,␤                                                                               |
|         |     title STRING NULL,␤                                                                              |
|         |     summary STRING NULL,␤                                                                            |
|         |     content STRING NULL,␤                                                                            |
|         |     link STRING NULL,␤                                                                               |
|         |     read BOOL NULL,␤                                                                                 |
|         |     date TIMESTAMP WITH TIME ZONE NULL,␤                                                             |
|         |     retrieved TIMESTAMP WITH TIME ZONE NULL,␤                                                        |
|         |     CONSTRAINT "primary" PRIMARY KEY (folder ASC, feed ASC, id ASC),␤                                |
|         |     UNIQUE INDEX article_id_key (id ASC),␤                                                           |
|         |     UNIQUE INDEX article_hash_key (hash ASC),␤                                                       |
|         |     FAMILY "primary" (id, feed, folder, hash, title, summary, content, link, read, date, retrieved)␤ |
|         | ) INTERLEAVE IN PARENT feed (folder, feed)                                                           |
+---------+------------------------------------------------------------------------------------------------------+
(1 row)
```

I performed the following query against this table:

```sql
SELECT id FROM article WHERE NOT read AND id > 0 limit 50;
```

- What did you expect to see?

I expected to get a nonzero number of results (since there are rows for which `read` is false). This is confirmed by doing the query without the `limit` clause:

```sql
root@:26257/Goliath> SELECT id FROM article WHERE NOT read AND id > 0;
+--------------------+
|         id         |
+--------------------+
| 250885803841617921 |
| 250889741747650561 |
| 250895619105882113 |
| 250917275225554945 |
| 250923144155168769 |
| 250925111932157953 |
| 250925111980163073 |
| 250934959014903809 |
| 250936942286602241 |
| 250940842501373953 |
| 250944774940196865 |
| 250950694693732353 |
+--------------------+
(12 rows)

```

- What did you see instead?

I instead get no returned rows when the `limit` parameter is small:

```sql
root@:26257/Goliath> SELECT id FROM article WHERE NOT read AND id > 0 limit 50;
+----+
| id |
+----+
+----+
(0 rows)

root@:26257/Goliath> SELECT id FROM article WHERE NOT read AND id > 0 limit 500;
+----+
| id |
+----+
+----+
(0 rows)
```
However, when the limit parameter is close to the number of rows in the table, I get results:

```sql
root@:26257/Goliath> select count(*) from article;
+----------+
| count(*) |
+----------+
|      699 |
+----------+
(1 row)

root@:26257/Goliath> SELECT id FROM article WHERE NOT read AND id > 0 limit 690;
+--------------------+
|         id         |
+--------------------+
| 250885803841617921 |
| 250889741747650561 |
| 250895619105882113 |
| 250917275225554945 |
| 250923144155168769 |
| 250925111932157953 |
| 250925111980163073 |
| 250934959014903809 |
+--------------------+
(8 rows)

root@:26257/Goliath> SELECT id FROM article WHERE NOT read AND id > 0 limit 698;
+--------------------+
|         id         |
+--------------------+
| 250885803841617921 |
| 250889741747650561 |
| 250895619105882113 |
| 250917275225554945 |
| 250923144155168769 |
| 250925111932157953 |
| 250925111980163073 |
| 250934959014903809 |
| 250936942286602241 |
| 250940842501373953 |
| 250944774940196865 |
+--------------------+
(11 rows)

root@:26257/Goliath> SELECT id FROM article WHERE NOT read AND id > 0 limit 699;
+--------------------+
|         id         |
+--------------------+
| 250885803841617921 |
| 250889741747650561 |
| 250895619105882113 |
| 250917275225554945 |
| 250923144155168769 |
| 250925111932157953 |
| 250925111980163073 |
| 250934959014903809 |
| 250936942286602241 |
| 250940842501373953 |
| 250944774940196865 |
| 250950694693732353 |
+--------------------+
(12 rows)

``` 

I was able to reproduce my expected results against a previous (<1.0) version of cockroachdb. I can try and figure out which version that was if it would be helpful.

**Comment by jrupac** [comment]
There's also unexpected truncation of returned rows in some cases too:

Here's everything:

```sql
root@:26257/Goliath> SELECT id FROM article WHERE NOT read order by id;
+--------------------+
|         id         |
+--------------------+
| 250885803841617921 |
| 250889741747650561 |
| 250895619105882113 |
| 250917275225554945 |
| 250923144155168769 |
| 250925111932157953 |
| 250925111980163073 |
| 250934959014903809 |
| 250936942286602241 |
| 250940842501373953 |
| 250944774940196865 |
| 250950694693732353 |
| 250958575645818881 |
| 250962490978598913 |
+--------------------+
(14 rows)
```

And here's some queries that yield partial results (the first is correct - there should be 13 rows returned).

```sql
root@:26257/Goliath> SELECT id FROM article WHERE NOT read and id > 250885803841617921 order by id;
+--------------------+
|         id         |
+--------------------+
| 250889741747650561 |
| 250895619105882113 |
| 250917275225554945 |
| 250923144155168769 |
| 250925111932157953 |
| 250925111980163073 |
| 250934959014903809 |
| 250936942286602241 |
| 250940842501373953 |
| 250944774940196865 |
| 250950694693732353 |
| 250958575645818881 |
| 250962490978598913 |
+--------------------+
(13 rows)

root@:26257/Goliath> SELECT id FROM article WHERE NOT read and id > 250885803841617921 limit 50;
+--------------------+
|         id         |
+--------------------+
| 250889741747650561 |
| 250895619105882113 |
| 250917275225554945 |
| 250923144155168769 |
| 250925111932157953 |
| 250925111980163073 |
| 250934959014903809 |
| 250936942286602241 |
+--------------------+
(8 rows)
```

**Comment by jordanlewis** [comment]
Thanks for the bug report! If you experienced this problem only after the 1.0 release, it's possible that you've discovered a bug in DistSQL. It's also possible that this will be fixed in an upcoming release by some of the changes introduced to fix similar bugs like #16163.

Please try running the queries with DistSQL disabled in your session (via `SET DISTSQL = OFF`) - that will help us figure out what's going on.

cc @andreimatei

**Comment by jrupac** [comment]
Hm, that did not seem to make any difference.

```sql
root@:26257/Goliath> SELECT id FROM article WHERE NOT read and id > 0 order by id;
+--------------------+
|         id         |
+--------------------+
| 250984099886235649 |
| 250988075467898881 |
[...snip...]
| 251023474624233473 |
+--------------------+
(38 rows)

root@:26257/Goliath> SELECT id FROM article WHERE NOT read and id > 0 order by id limit 50;
+----+
| id |
+----+
+----+
(0 rows)

root@:26257/Goliath> set distsql = off;
SET

root@:26257/Goliath> SELECT id FROM article WHERE NOT read and id > 0 order by id;
+--------------------+
|         id         |
+--------------------+
| 250984099886235649 |
| 250988075467898881 |
[...snip...]
| 251023474624233473 |
+--------------------+
(38 rows)

root@:26257/Goliath> SELECT id FROM article WHERE NOT read and id > 0 order by id limit 50;
+----+
| id |
+----+
+----+
(0 rows)
```

**Comment by knz** [comment]
cc @RaduBerinde can you have a look?

**Comment by RaduBerinde** [comment]
```
root@:26257/test> EXPLAIN (VERBOSE) SELECT id FROM article WHERE NOT read AND id > 0 limit 50;
+-------+------------+----------+------------------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------+----------+
| Level |    Type    |  Field   |      Description       |                                                                              Columns                                                                              | Ordering |
+-------+------------+----------+------------------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------+----------+
|     0 | limit      |          |                        | (id)                                                                                                                                                              |          |
|     0 |            | count    |                     50 |                                                                                                                                                                   |          |
|     1 | render     |          |                        | (id)                                                                                                                                                              |          |
|     1 |            | render 0 | test.article.id        |                                                                                                                                                                   |          |
|     2 | index-join |          |                        | (id, feed[omitted], folder[omitted], hash[omitted], title[omitted], summary[omitted], content[omitted], link[omitted], read, "date"[omitted], retrieved[omitted]) |          |
|     3 | scan       |          |                        | (id, feed, folder, hash[omitted], title[omitted], summary[omitted], content[omitted], link[omitted], read[omitted], "date"[omitted], retrieved[omitted])          |          |
|     3 |            | table    | article@article_id_key |                                                                                                                                                                   |          |
|     3 |            | spans    | /1-                    |                                                                                                                                                                   |          |
|     3 |            | limit    |                     50 |                                                                                                                                                                   |          |
|     3 | scan       |          |                        | (id, feed[omitted], folder[omitted], hash[omitted], title[omitted], summary[omitted], content[omitted], link[omitted], read, "date"[omitted], retrieved[omitted]) |          |
|     3 |            | table    | article@primary        |                                                                                                                                                                   |          |
|     3 |            | filter   | NOT read               |                                                                                                                                                                   |          |
+-------+------------+----------+------------------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------+----------+
```

We have a hard limit at the index join level. This becomes a hard limit at the index scan level which is incorrect - we have a filter that gets applied after we join with the table.

In `applyLimit` we have:
```go
	case *indexJoinNode:
		applyLimit(n.index, numRows, soft)
		setUnlimited(n.table)
```
We need to make this always be soft if `indexJoinNode.table` has a filter. I can make the change.

**Comment by RaduBerinde** [comment]
Note that even though this bug was in there for a while, I believe it only leads to incorrect results since #15925.

**Comment by knz** [comment]
yes that seems correct... That's a schroeninbug that keeps on giving.

**Comment by RaduBerinde** [comment]
@jrupac - sorry, I meant to send this 2 days ago. The fix will be in 1.0.2. In the meantime, as a workaround you can make sure that the index includes the `read` column: `UNIQUE INDEX article_id_key (id ASC) STORING (read)`. This will also make the query more efficient.

You can also try reverting to 1.0 which I think doesn't exhibit this problem.

**Comment by jrupac** [comment]
Thanks all for super quick fix and thanks @RaduBerinde for the advice.

**Comment by jrupac** [comment]
Not that it matters much, but it didn't seem like even adding a storing index helped the issue with v1.0.1. I ended up reverting back to v1.0 and that correctly did not exhibit this problem.

**Comment by jrupac** [comment]
I've added a new storing index as follows:

```sql
root@:26257/goliath> show create table article;
+---------+------------------------------------------------------------------------------------------------------+
|  Table  |                                             CreateTable                                              |
+---------+------------------------------------------------------------------------------------------------------+
| article | CREATE TABLE article (␤                                                                              |
|         |     id INT NOT NULL DEFAULT unique_rowid(),␤                                                         |
|         |     feed INT NOT NULL,␤                                                                              |
|         |     folder INT NOT NULL,␤                                                                            |
|         |     hash STRING NULL,␤                                                                               |
|         |     title STRING NULL,␤                                                                              |
|         |     summary STRING NULL,␤                                                                            |
|         |     content STRING NULL,␤                                                                            |
|         |     link STRING NULL,␤                                                                               |
|         |     read BOOL NULL,␤                                                                                 |
|         |     date TIMESTAMP WITH TIME ZONE NULL,␤                                                             |
|         |     retrieved TIMESTAMP WITH TIME ZONE NULL,␤                                                        |
|         |     CONSTRAINT "primary" PRIMARY KEY (folder ASC, feed ASC, id ASC),␤                                |
|         |     UNIQUE INDEX article_id_key (id ASC),␤                                                           |
|         |     UNIQUE INDEX article_hash_key (hash ASC),␤                                                       |
|         |     UNIQUE INDEX article_idx_read_key (id ASC) STORING (read),␤                                      |
|         |     FAMILY "primary" (id, feed, folder, hash, title, summary, content, link, read, date, retrieved)␤ |
|         | ) INTERLEAVE IN PARENT feed (folder, feed)                                                           |
+---------+------------------------------------------------------------------------------------------------------+
(1 row)
```

(I had to create a new index on the table instead of replacing the existing "article_id_key" since I think that's created automatically from "id" being a unique primary key. I don't think this is problematic.)

Ah, I think I understand part of the problem. I had simplified the query for this bug; the actual query that I'm performing that's returning no results is:

```sql
root@:26257/goliath> SELECT id, feed, folder, title, summary, content, link, date FROM article WHERE NOT read and id > 0 order by id limit 50;
+----+------+--------+-------+---------+---------+------+------+
| id | feed | folder | title | summary | content | link | date |
+----+------+--------+-------+---------+---------+------+------+
+----+------+--------+-------+---------+---------+------+------+
(0 rows)
```

which looks like it doesn't use the "storing" index:

```sql
root@:26257/goliath> explain(verbose) SELECT id, feed, folder, title, summary, content, link, date FROM article WHERE NOT read and id > 0 order by id limit 50;
+-------+------------+----------+-------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------+------------+
| Level |    Type    |  Field   |       Description       |                                                                        Columns                                                                         |  Ordering  |
+-------+------------+----------+-------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------+------------+
|     0 | limit      |          |                         | (id, feed, folder, title, summary, content, link, date)                                                                                                | +id,unique |
|     0 |            | count    |                      50 |                                                                                                                                                        |            |
|     1 | render     |          |                         | (id, feed, folder, title, summary, content, link, date)                                                                                                | +id,unique |
|     1 |            | render 0 | goliath.article.id      |                                                                                                                                                        |            |
|     1 |            | render 1 | goliath.article.feed    |                                                                                                                                                        |            |
|     1 |            | render 2 | goliath.article.folder  |                                                                                                                                                        |            |
|     1 |            | render 3 | goliath.article.title   |                                                                                                                                                        |            |
|     1 |            | render 4 | goliath.article.summary |                                                                                                                                                        |            |
|     1 |            | render 5 | goliath.article.content |                                                                                                                                                        |            |
|     1 |            | render 6 | goliath.article.link    |                                                                                                                                                        |            |
|     1 |            | render 7 | goliath.article.date    |                                                                                                                                                        |            |
|     2 | index-join |          |                         | (id, feed, folder, hash[omitted], title, summary, content, link, read, date, retrieved[omitted])                                                       | +id,unique |
|     3 | scan       |          |                         | (id, feed, folder, hash[omitted], title[omitted], summary[omitted], content[omitted], link[omitted], read[omitted], date[omitted], retrieved[omitted]) | +id,unique |
|     3 |            | table    | article@article_id_key  |                                                                                                                                                        |            |
|     3 |            | spans    | /1-                     |                                                                                                                                                        |            |
|     3 |            | limit    |                      50 |                                                                                                                                                        |            |
|     3 | scan       |          |                         | (id, feed, folder, hash[omitted], title, summary, content, link, read, date, retrieved[omitted])                                                       |            |
|     3 |            | table    | article@primary         |                                                                                                                                                        |            |
|     3 |            | filter   | NOT read                |                                                                                                                                                        |            |
+-------+------------+----------+-------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------+------------+
(19 rows)
```
I think that makes sense since the storing index doesn't have the rest of the columns I'm querying. 

However, even forcing the use of the new storing index yields no results - this I'm not sure is WAI or not:

```sql
root@:26257/goliath> SELECT id, feed, folder, title, summary, content, link, date FROM article@article_idx_read_key WHERE NOT read and id > 0 order by id limit 50;
+----+------+--------+-------+---------+---------+------+------+
| id | feed | folder | title | summary | content | link | date |
+----+------+--------+-------+---------+---------+------+------+
+----+------+--------+-------+---------+---------+------+------+
(0 rows)
root@:26257/goliath> explain(verbose) SELECT id, feed, folder, title, summary, content, link, date FROM article@article_idx_read_key WHERE NOT read and id > 0 order by id limit 50;
+-------+------------+----------+------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------+------------+
| Level |    Type    |  Field   |         Description          |                                                                        Columns                                                                         |  Ordering  |
+-------+------------+----------+------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------+------------+
|     0 | limit      |          |                              | (id, feed, folder, title, summary, content, link, date)                                                                                                | +id,unique |
|     0 |            | count    |                           50 |                                                                                                                                                        |            |
|     1 | render     |          |                              | (id, feed, folder, title, summary, content, link, date)                                                                                                | +id,unique |
|     1 |            | render 0 | goliath.article.id           |                                                                                                                                                        |            |
|     1 |            | render 1 | goliath.article.feed         |                                                                                                                                                        |            |
|     1 |            | render 2 | goliath.article.folder       |                                                                                                                                                        |            |
|     1 |            | render 3 | goliath.article.title        |                                                                                                                                                        |            |
|     1 |            | render 4 | goliath.article.summary      |                                                                                                                                                        |            |
|     1 |            | render 5 | goliath.article.content      |                                                                                                                                                        |            |
|     1 |            | render 6 | goliath.article.link         |                                                                                                                                                        |            |
|     1 |            | render 7 | goliath.article.date         |                                                                                                                                                        |            |
|     2 | index-join |          |                              | (id, feed, folder, hash[omitted], title, summary, content, link, read, date, retrieved[omitted])                                                       | +id,unique |
|     3 | scan       |          |                              | (id, feed, folder, hash[omitted], title[omitted], summary[omitted], content[omitted], link[omitted], read[omitted], date[omitted], retrieved[omitted]) | +id,unique |
|     3 |            | table    | article@article_idx_read_key |                                                                                                                                                        |            |
|     3 |            | spans    | /1-                          |                                                                                                                                                        |            |
|     3 |            | limit    |                           50 |                                                                                                                                                        |            |
|     3 | scan       |          |                              | (id, feed, folder, hash[omitted], title, summary, content, link, read, date, retrieved[omitted])                                                       |            |
|     3 |            | table    | article@primary              |                                                                                                                                                        |            |
|     3 |            | filter   | NOT read                     |                                                                                                                                                        |            |
+-------+------------+----------+------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------+------------+
(19 rows)
```
---

Just querying for "id" and "read" works with the new index:

```sql
root@:26257/goliath>  SELECT id, read FROM article WHERE NOT read and id > 0 order by id limit 5;
+--------------------+-------+
|         id         | read  |
+--------------------+-------+
| 252104345187811329 | false |
| 252106347292590081 | false |
| 252108273880694785 | false |
| 252108277940649985 | false |
| 252116138799300609 | false |
+--------------------+-------+
(5 rows)
root@:26257/goliath> explain(verbose) SELECT id, read FROM article WHERE NOT read and id > 0 order by id limit 5;
+-------+--------+----------+------------------------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------+------------+
| Level |  Type  |  Field   |         Description          |                                                                             Columns                                                                             |  Ordering  |
+-------+--------+----------+------------------------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------+------------+
|     0 | limit  |          |                              | (id, read)                                                                                                                                                      | +id,unique |
|     0 |        | count    |                            5 |                                                                                                                                                                 |            |
|     1 | render |          |                              | (id, read)                                                                                                                                                      | +id,unique |
|     1 |        | render 0 | goliath.article.id           |                                                                                                                                                                 |            |
|     1 |        | render 1 | goliath.article.read         |                                                                                                                                                                 |            |
|     2 | scan   |          |                              | (id, feed[omitted], folder[omitted], hash[omitted], title[omitted], summary[omitted], content[omitted], link[omitted], read, date[omitted], retrieved[omitted]) | +id,unique |
|     2 |        | table    | article@article_idx_read_key |                                                                                                                                                                 |            |
|     2 |        | spans    | /1-                          |                                                                                                                                                                 |            |
|     2 |        | filter   | NOT read                     |                                                                                                                                                                 |            |
+-------+--------+----------+------------------------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------+------------+
(9 rows)
```

**TL;DR**: My actual query is (correctly) failing to use the new storing index. Unclear to me if even forcing the use of the index should still not yield any results.

If I wanted to correctly fix this, I guess I should just create a storing index with everything I need to retrieve, right?

Sorry for the noise if this part ends up being WAI.

**Comment by RaduBerinde** [comment]
Thanks for following up, this is helpful!

Yes, it seems that the index is not being used because we need an index-join anyway. And as far as the planning process is concerned, the two indexes are the same so it chooses the "smaller" index (fewer columns). But in this case, the planning process should take into account that it's more efficient if the entire filter is applied on the index. I filed #16426 to track this improvement.

For your particular case, I would just drop the `article_id_key` index (it's not useful if we also have `article_idx_read_key`, it just adds overhead on insertions).
